File: web/src/lib/__tests__/format.test.ts
import { describe, it, expect } from 'vitest';
import { formatMoney, formatPct, formatSignedMoney, trendClass, formatQty } from '../format';

describe('formatMoney INR (Indian grouping)', () => {
  it('groups lakhs + crores correctly', () => {
    expect(formatMoney('4422540.74')).toBe('44,22,540.74');
    expect(formatMoney('6288819.43')).toBe('62,88,819.43');
  });
  it('precision is preserved at large magnitudes (no Number coercion loss)', () => {
    expect(formatMoney('123456789012345.67')).toBe(
      '12,34,56,78,90,12,345.67',
    );
  });
  it('two fractional digits always shown', () => {
    expect(formatMoney('100')).toBe('100.00');
  });
});

describe('formatPct', () => {
  it('decimal fraction → human percent with sign', () => {
    expect(formatPct('0.0421')).toBe('+4.21%');
    expect(formatPct('-0.0095')).toBe('-0.95%');
  });
  it('zero is unsigned', () => {
    expect(formatPct('0')).toBe('0.00%');
  });
});

describe('formatSignedMoney', () => {
  it('+ for positive, - for negative', () => {
    expect(formatSignedMoney('15190.51')).toBe('+15,190.51');
    expect(formatSignedMoney('-432.04')).toBe('-432.04');
  });
});

describe('trendClass', () => {
  it('returns gain/loss/flat colour classes', () => {
    expect(trendClass('1')).toBe('text-gain');
    expect(trendClass('-1')).toBe('text-loss');
    expect(trendClass('0')).toBe('text-flat');
  });
});

describe('formatQty', () => {
  it('strips trailing decimal zeros', () => {
    expect(formatQty('10.0000')).toBe('10');
    expect(formatQty('10.5000')).toBe('10.5');
  });
});

File: android/core/src/test/kotlin/com/tickernest/core/FormatTest.kt
package com.tickernest.core

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class FormatTest {

    @Test fun `INR groups lakhs and crores`() {
        assertEquals("44,22,540.74", formatMoney(D("4422540.74")))
        assertEquals("62,88,819.43", formatMoney(D("6288819.43")))
    }

    @Test fun `INR precision preserved at very large magnitudes`() {
        assertEquals(
            "12,34,56,78,90,12,345.67",
            formatMoney(D("123456789012345.67")),
        )
    }

    @Test fun `USD uses thousands grouping`() {
        assertEquals("1,234,567.89", formatMoney(D("1234567.89"), Currency.USD))
    }

    @Test fun `formatPct adds a sign for non-zero values`() {
        assertEquals("+4.21%", formatPct(D("0.0421")))
        assertEquals("-0.95%", formatPct(D("-0.0095")))
        assertEquals("0.00%", formatPct(D("0")))
    }

    @Test fun `formatSignedMoney prefixes plus for positive and minus for negative`() {
        assertEquals("+15,190.51", formatSignedMoney(D("15190.51")))
        assertEquals("-432.04", formatSignedMoney(D("-432.04")))
        assertEquals("0.00", formatSignedMoney(D("0")))
    }

    @Test fun `formatQty strips trailing zeros and the decimal point`() {
        assertEquals("10", formatQty(D("10.0000")))
        assertEquals("10.5", formatQty(D("10.5000")))
        assertEquals("10.5001", formatQty(D("10.5001")))
    }
}
